File: app/src/main/java/com/emiryanvl/tapreader/ui/viewModels/LibraryViewModel.java
package com.emiryanvl.tapreader.ui.viewModels;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.emiryanvl.tapreader.domain.model.Book;
import com.emiryanvl.tapreader.domain.usecases.GetBooksBySubjectUseCase;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class LibraryViewModel extends ViewModel {

    private final GetBooksBySubjectUseCase getBooksBySubjectUseCase;
    public MutableLiveData<List<Book>> subjectBooks = new MutableLiveData<>();

    @Inject
    public LibraryViewModel(
        GetBooksBySubjectUseCase getBooksBySubjectUseCase
    ) {
        this.getBooksBySubjectUseCase = getBooksBySubjectUseCase;
    }

    public void getBooksBySubject() {
        Disposable disposable = getBooksBySubjectUseCase.execute("fiction")
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(sb -> {
                subjectBooks.setValue(sb);
            });
    }
}

File: app/src/main/java/com/emiryanvl/tapreader/ui/viewModels/LibraryViewModel.kt
package com.emiryanvl.tapreader.ui.viewModels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.emiryanvl.tapreader.domain.models.Book
import com.emiryanvl.tapreader.domain.usecases.GetQueryBooksUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class LibraryViewModel @Inject constructor(
    private val getFilteredBooksUseCase: GetQueryBooksUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(UiState())
    val uiState = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            getFilteredBooksUseCase("Михаил Елизаров").collect {
                _uiState.value = _uiState.value.copy(bookList = it)
            }
        }
    }

    data class UiState(
        val bookList: List<Book> = emptyList()
    )
}